### ISSUE TITLE
No normalization for LinearRegression causes infinite values

### ISSUE THREAD

**Issue by mjovanc** [blackportal-ai/delta#248]
In `delta_examples/classical/linear_regression`.

The `LinearRegression` implementation in `deltaml` fails with `NaN` predictions due to numerical instability caused by the lack of data normalization, as observed when training on the `California Housing dataset`. Without normalizing features (`x_train`) and target (`y_train`), which have vastly different scales (e.g., `median_house_value` ranges from `15,000` to `500,000`, while `median_income` is `0–15`), the initial predictions and errors (`predictions - y`) become extremely large. This leads to massive gradients (e.g., `-1160130361.8696012` in `epoch 0`), which, even with a small learning rate (`0.00001`) and clipping (`±0.01`), accumulate over epochs, causing weights to grow exponentially (loss jumps from `5.58e10` to `3.75e296` by `epoch 26`, then `inf` and `NaN` by `epoch 55`). The absence of built-in normalization in the library exacerbates this, as unscaled data overwhelms gradient descent, pushing values beyond `f64`’s representable range.

```sh
   Compiling linear_regression v0.1.0 (/Users/k01522/Private/delta/delta_examples/classical/linear_regression)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1.44s
     Running `/Users/k01522/Private/delta/target/debug/linear_regression`
x_train shape: (14448, 8)
Epoch 0: Loss = 55788267509.04506
Epoch 0: Loss = 55788267509.04506, Weights = Some([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], shape=[8], strides=[1], layout=CFcf (0xf), const ndim=1), Bias = 0.0, GradW = [49322512.61312019, -14622472.759260816, -12149662.02893134, -1160130361.8696012, -225871360.0444352, -583555169.3794297, -212912098.0685216, -1894136.9422600165], shape=[8], strides=[1], layout=CFcf (0xf), const ndim=1, GradB = -412315.66708194907
Epoch 1: Loss = 2.838399021423067e21
Epoch 2: Loss = 2.870436286473804e32
Epoch 3: Loss = 2.902836453245759e43
Epoch 4: Loss = 2.9356023385264217e54
Epoch 5: Loss = 2.9687380700784045e65
Epoch 6: Loss = 3.002247822556558e76
Epoch 7: Loss = 3.036135817737384e87
Epoch 8: Loss = 3.070406325050856e98
Epoch 9: Loss = 3.1050636621183393e109
Epoch 10: Loss = 3.140112195296515e120
Epoch 11: Loss = 3.1755563402274767e131
Epoch 12: Loss = 3.2114005623951157e142
Epoch 13: Loss = 3.247649377687584e153
Epoch 14: Loss = 3.2843073529663435e164
Epoch 15: Loss = 3.321379106641517e175
Epoch 16: Loss = 3.358869309253679e186
Epoch 17: Loss = 3.396782684062309e197
Epoch 18: Loss = 3.435124007640876e208
Epoch 19: Loss = 3.4738981104787047e219
Epoch 20: Loss = 3.5131098775893604e230
Epoch 21: Loss = 3.552764249126219e241
Epoch 22: Loss = 3.5928662210048925e252
Epoch 23: Loss = 3.6334208455325217e263
Epoch 24: Loss = 3.6744332320444686e274
Epoch 25: Loss = 3.7159085475477964e285
Epoch 26: Loss = 3.757852017372527e296
Epoch 27: Loss = inf
Epoch 28: Loss = inf
Epoch 29: Loss = inf
Epoch 30: Loss = inf
Epoch 31: Loss = inf
Epoch 32: Loss = inf
Epoch 33: Loss = inf
Epoch 34: Loss = inf
Epoch 35: Loss = inf
Epoch 36: Loss = inf
Epoch 37: Loss = inf
Epoch 38: Loss = inf
Epoch 39: Loss = inf
Epoch 40: Loss = inf
Epoch 41: Loss = inf
Epoch 42: Loss = inf
Epoch 43: Loss = inf
Epoch 44: Loss = inf
Epoch 45: Loss = inf
Epoch 46: Loss = inf
Epoch 47: Loss = inf
Epoch 48: Loss = inf
Epoch 49: Loss = inf
Epoch 50: Loss = inf
Epoch 51: Loss = inf
Epoch 52: Loss = inf
Epoch 53: Loss = inf
Epoch 54: Loss = inf
Epoch 55: Loss = NaN
NaN detected at epoch 55
Predictions for test data: [NaN, NaN, NaN, NaN, NaN, ..., NaN, NaN, NaN, NaN, NaN], shape=[6192], strides=[1], layout=CFcf (0xf), const ndim=1
Test Loss (MSE) after training: NaN
```

**Comment by mjovanc** [comment]
Solved.
